specs/commandline/verify.md

```diff
@@ -37,14 +52,16 @@ Usage:
 Flags:
        --allow-referrers-api         [Experimental] use the Referrers API to verify signatures, if not supported (returns 404), fallback to the Referrers tag schema
   -d,  --debug                       debug mode
+       --file                        enable verifying a file, if set, the reference argument is the file path or full URI reference of the file artifact in registry (required if --signature is set)
   -h,  --help                        help for verify
        --insecure-registry           use HTTP protocol while connecting to registries. Should be used only for testing
        --max-signatures int          maximum number of signatures to evaluate or examine (default 100)
        --oci-layout                  [Experimental] verify the artifact stored as OCI image layout
   -p,  --password string             password for registry operations (default to $NOTATION_PASSWORD if not specified)
        --plugin-config stringArray   {key}={value} pairs that are passed as it is to a plugin, if the verification is associated with a verification plugin, refer plugin documentation to set appropriate values
-       --scope string                [Experimental] set trust policy scope for artifact verification, required and can only be used when flag "--oci-layout" is set
+       --scope string                [Experimental] set trust policy scope for artifact verification, required when flag "--oci-layout" is set, can only be used when "--oci-layout" or "--file" is set
```

Comment:
@priteshbandi This is an existing flag. It was initially introduced for verifying artifact as `oci-layout`. In verifying a `file` scenario, it's only needed when the file and the signature are in user's file system. In this case, the user has two choices, they are mentioned in this PR (verify.md) as well, please take a look there.

Reply:
Yes but `registryScope` doesn't mean anything if user is verifying a blob/file.

Also as per https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#registry-scopes-constraints `List of one or more fully qualified repository URIs. The repository URI MUST NOT contain the asterisk character *.`

Reply:
@priteshbandi Agree. Yes, we should have a spec change in the https://github.com/notaryproject/specifications repo as well especially for the `trust policy` part.

Reply:
@priteshbandi I will add the related spec changes into the specification's repo. Here's the suggestion from @yizha1 on the spec changes: https://github.com/notaryproject/notation/issues/741#issuecomment-1686351235

Reply:
@priteshbandi Answering them below:
> This will also require updates to [descriptor](https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md#payload) which is being signing.

To be more precise, this is the update to the signature `payload`. 
And yes, I'm working on the spec in the [specifications](https://github.com/notaryproject/specifications) repo to introduce non-OCI-signature-specifications which includes `signature payload`, `signature storage`, `trust policy`, and any change on `sign and verification workflow`. This work is in progress, and I will create the PR once it's ready. 

> ```
> {
>   "targetArtifact": {
>     "mediaType": "sbom/example", 
>     "digest": "sha256:9834876dcfb05cb167a5c24953eba58c4ac89b1adf57f28f2f9d09af107ee8f0",
>     "size": 32654 <-- 
>   }
> }
> ```
> 
> * What would be the media type ?

Good point. We should allow users to pick their own media type. We could do this by:
`notation sign --file myFile:sbom/example`
i.e., using ':' to delimit the file path and its media type. What do you think?
/cc: @shizhMSFT 

> * Should we give control to user on which hashing algo to use in `digest` calculation?

Currently, I'm using [opencontainers/go-digest](https://pkg.go.dev/github.com/opencontainers/go-digest@v1.0.0#FromReader) to compute the digest. The algorithm is SHA256, which is the primary storage digest (https://pkg.go.dev/github.com/opencontainers/go-digest#Canonical).

> * `size`, is this the size of actual content that we are trying to sign?

Yes, since the actual content is signed, the size corresponds to the size of the actual content.

> ##### Signing process look like this?
> 1. Notation Reads the content that needs be signed (can be read from file or stdin or flag value)
> 2. Notation creates descriptor from the content, which needs following operation
> 3. Calculate hash/digest and size of content.
> 4. Calculate hash/digest of descriptor. This part will be either done by plugin or by notation depending upon plugin type.
> 5. Sign the hash/digest of descriptor. This would be done by plugin
> 6. return detached signature to user.
> 
> _Here we are calculating digest twice instead of once (ideal). Should we worry about this?_ Thinking more we will probably need this to support user defined metadata.

I don't think we should worry about this? Because even for the current Notation, we create the OCI descriptor out of the target OCI artifact, then sign/verify the descriptor.

> ##### For verification
> 1. Notation reads the detached signature
> 2. Verify the signature is valid (authenticity, integrity, expiry, etc checks)
> 3. Read the descriptor from signature.
> 4. Notation Reads the content to be verified
> 5. Calculate hash/digest of content and its size
> 6. Verify that ash/digest of content and its size matches with the descriptor stored in signature.

Yes, it reuses what we have in the current Notation workflow. The changes for sign and verify are actually small in our code base.

Reply:
> > * What would be the media type ?
> 
> Good point. We should allow users to pick their own media type. We could do this by: `notation sign --file myFile:sbom/example` i.e., using ':' to delimit the file path and its media type. We should have a default if user doesn't pass their own media type, we should define this default in the specifications repo though. What do you think? /cc: @shizhMSFT
> Good point. We should allow users to pick their own media type. 
 
IMO, for future extensibility and robustness, we should control the media type.



> > * Should we give control to user on which hashing algo to use in `digest` calculation?
> 
> Currently, I'm using [opencontainers/go-digest](https://pkg.go.dev/github.com/opencontainers/go-digest@v1.0.0#FromReader) to compute the digest. The algorithm is SHA256, which is the primary storage digest (https://pkg.go.dev/github.com/opencontainers/go-digest#Canonical).

 IMO, we can decide for sane default but allow user to control hashing algo. Depending upon usecase and compliance need user might want to use different hashing algos.

> > ##### Signing process look like this?
> > 
> > 1. Notation Reads the content that needs be signed (can be read from file or stdin or flag value)
> > 2. Notation creates descriptor from the content, which needs following operation
> > 3. Calculate hash/digest and size of content.
> > 4. Calculate hash/digest of descriptor. This part will be either done by plugin or by notation depending upon plugin type.
> > 5. Sign the hash/digest of descriptor. This would be done by plugin
> > 6. return detached signature to user.
> > 
> > _Here we are calculating digest twice instead of once (ideal). Should we worry about this?_ Thinking more we will probably need this to support user defined metadata.
> 
> I don't think we should worry about this? Because even for the current Notation, we create the OCI descriptor out of the target OCI artifact, then sign/verify the descriptor.
True, we could have directly signed pre-calculated digest(without rehashing). Lets discuss this in meeting

Reply:
@priteshbandi Since this PR is closed, could you move the discussions to https://github.com/notaryproject/notation/discussions/767?